File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/interfaces/ApiErrors.java
package dev.vra.poc00.interfaces;

import dev.vra.poc00.domain.DomainFailure;
import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatusCode;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.context.request.ServletWebRequest;

@RestControllerAdvice
public class ApiErrors extends ResponseEntityExceptionHandler {
    public record Error(String code, int status, String requestId, String message) {}

    @ExceptionHandler(DomainFailure.class)
    ResponseEntity<Object> domain(DomainFailure failure, HttpServletRequest request) {
        return switch (failure) {
            case DomainFailure.InsufficientStock ignored -> error(409, "inventory.insufficient_stock", "Insufficient stock.", request);
            case DomainFailure.UnknownSku ignored -> error(404, "inventory.unknown_sku", "SKU was not found.", request);
            case DomainFailure.InvalidQuantity ignored -> error(400, "request.invalid", "Quantity must be positive.", request);
            case DomainFailure.InvalidTransition ignored -> error(409, "order.invalid_transition", "Order transition is not allowed.", request);
            case DomainFailure.InvalidIdempotencyKey ignored -> error(400, "request.invalid_idempotency_key", "Invalid idempotency key.", request);
            case DomainFailure.CurrencyMismatch ignored -> error(422, "money.currency_mismatch", "Currencies must match.", request);
        };
    }

    @Override
    protected ResponseEntity<Object> handleExceptionInternal(Exception exception, Object body,
            HttpHeaders headers, HttpStatusCode status, WebRequest request) {
        var servletRequest = ((ServletWebRequest) request).getRequest();
        var safe = error(status.value(), status.is4xxClientError() ? "request.invalid" : "internal.error",
                status.is4xxClientError() ? "Invalid request." : "Internal server error.", servletRequest);
        return new ResponseEntity<>(safe.getBody(), headers, status);
    }

    @ExceptionHandler(Exception.class)
    ResponseEntity<Object> unexpected(Exception exception, HttpServletRequest request) {
        // Deliberately omit exception message/stack: driver messages may contain sensitive data.
        LoggerFactory.getLogger(ApiErrors.class).error("event=internal_error");
        return error(500, "internal.error", "Internal server error.", request);
    }

    private static ResponseEntity<Object> error(int status, String code, String message, HttpServletRequest request) {
        return ResponseEntity.status(status).body(new Error(code, status,
                (String) request.getAttribute(RequestIdFilter.ATTRIBUTE), message));
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/interfaces/ApiErrors.kt
package dev.vra.poc00.kotlin.interfaces

import dev.vra.poc00.kotlin.domain.DomainFailure
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.ServletWebRequest
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@RestControllerAdvice
class ApiErrors : ResponseEntityExceptionHandler() {
    data class Error(val code: String, val status: Int, val requestId: String?, val message: String)

    @ExceptionHandler(DomainFailure::class)
    fun domain(failure: DomainFailure, request: HttpServletRequest): ResponseEntity<Any> =
        when (failure) {
            is DomainFailure.InsufficientStock ->
                error(409, "inventory.insufficient_stock", "Insufficient stock.", request)
            is DomainFailure.UnknownSku ->
                error(404, "inventory.unknown_sku", "SKU was not found.", request)
            is DomainFailure.InvalidQuantity ->
                error(400, "request.invalid", "Quantity must be positive.", request)
            is DomainFailure.InvalidTransition ->
                error(409, "order.invalid_transition", "Order transition is not allowed.", request)
            is DomainFailure.InvalidIdempotencyKey ->
                error(400, "request.invalid_idempotency_key", "Invalid idempotency key.", request)
            is DomainFailure.CurrencyMismatch ->
                error(422, "money.currency_mismatch", "Currencies must match.", request)
            is DomainFailure.InvalidSkuId ->
                error(400, "request.invalid_sku_id", "Invalid SKU identifier.", request)
        }

    override fun handleExceptionInternal(
        ex: Exception,
        body: Any?,
        headers: HttpHeaders,
        statusCode: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {
        val servletRequest = (request as ServletWebRequest).request
        val safe = error(
            statusCode.value(),
            if (statusCode.is4xxClientError) "request.invalid" else "internal.error",
            if (statusCode.is4xxClientError) "Invalid request." else "Internal server error.",
            servletRequest
        )
        return ResponseEntity.status(statusCode).headers(headers).body<Any>(safe.body)
    }

    @ExceptionHandler(Exception::class)
    fun unexpected(exception: Exception, request: HttpServletRequest): ResponseEntity<Any> {
        // Driver text and stack traces can disclose details; deliberately omit them.
        apiLogger.error("event=internal_error")
        return error(500, "internal.error", "Internal server error.", request)
    }

    private fun error(status: Int, code: String, message: String, request: HttpServletRequest): ResponseEntity<Any> =
        ResponseEntity.status(status).body<Any>(
            Error(code, status, request.getAttribute(RequestIdFilter.ATTRIBUTE) as? String, message)
        )

    companion object {
        private val apiLogger = LoggerFactory.getLogger(ApiErrors::class.java)
    }
}
